### PR TITLE
@W-10778096@ Handle unresolvable parameters passed to database operations

### PR DESCRIPTION
**Pull request by rmohan20**
1. Added new test cases to confirm that MethodCallExpression passed to Database.<operation>() is not handled only when the method call cannot be resolved.
2. Created a new violation subtype, `UnresolvedCrudFlsViolation`, to be thrown when SFGE is unable to resolve the parameter passed in. This also covers BinaryExpression vertices cases we have in our backlog.
3. Separated violation creation and violation message creation.
4. Moved user-facing message templates to the common `UserFacingMessages` class.
5. Related test framework changes

### CODE REVIEW

File: sfge/src/main/java/com/salesforce/config/UserFacingMessages.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+public final class UserFacingMessages {
+
+    /** UserActionException * */
+
+    // format: filename,defined type, line number
+    public static final String UNREACHABLE_CODE =
+            "Please remove unreachable code to proceed with analysis: %s,%s:%d";
+
+    /** CRUD/FLS Violation messages * */
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/config/UserFacingMessages.java:15]: @teresa-allen-sfdc These messages could use your magic. These have existed for a while but putting them in one place for easier review.

File: sfge/src/main/java/com/salesforce/graph/build/MethodPathBuilderVisitor.java [hunk 2]
```diff
@@ -597,7 +597,7 @@ private void addEdge(String name, Vertex from, Vertex to) {
             // Ask user to fix unreachable code
             throw new UserActionException(
                     String.format(
-                            UserFacingErrorMessages.UNREACHABLE_CODE,
+                            UserFacingMessages.UNREACHABLE_CODE,
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/build/MethodPathBuilderVisitor.java:600]: Class rename has created some ripples.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexStringValue.java [hunk 3]
```diff
@@ -156,6 +157,9 @@ public TypeableUtil.OrderedTreeSet getTypes() {
                             Pair.of(METHOD_ESCAPE_HTML_3, StringEscapeUtils::escapeHtml3),
                             Pair.of(METHOD_ESCAPE_HTML_4, StringEscapeUtils::escapeHtml4),
                             Pair.of(METHOD_ESCAPE_JAVA, StringEscapeUtils::escapeJava),
+                            Pair.of(
+                                    METHOD_ESCAPE_SINGLE_QUOTES,
+                                    (stringValue) -> stringValue.replace("'", "\\'")),
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexStringValue.java:162]: Couldn't find a equivalent in Apache Commons

File: sfge/src/main/java/com/salesforce/rules/PathBasedRuleRunner.java [hunk 4]
```diff
@@ -125,10 +125,11 @@ private void convertToViolations(HashSet<FlsViolationInfo> flsViolationInfos) {
         // Consolidate/regroup FLS violations across paths so that there are no
         // duplicates with different field sets for the same source/sink/dmlOperation
         final HashSet<FlsViolationInfo> consolidatedFlsViolationInfos =
-                FlsViolationUtils.consolidateFlsViolations(flsViolationInfos);
+                FlsViolationMessageUtil.consolidateFlsViolations(flsViolationInfos);
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/rules/PathBasedRuleRunner.java:128]: Side effects from refactoring.

File: sfge/src/main/java/com/salesforce/rules/fls/apex/operations/FlsViolationMessageUtil.java [hunk 5]
```diff
@@ -3,35 +3,23 @@
...
 import java.util.HashSet;
 import java.util.List;
 import java.util.Optional;
-import java.util.Set;
 import java.util.TreeSet;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 /** Utils class that deals with FLS-specific violations */
-public final class FlsViolationUtils {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/rules/fls/apex/operations/FlsViolationMessageUtil.java:23]: Moved these out to UserFacingMessages.java

File: sfge/src/main/java/com/salesforce/rules/fls/apex/operations/FlsViolationMessageUtil.java [hunk 6]
```diff
@@ -228,66 +215,4 @@ private static boolean hasUnhandledSegment(String input) {
         return RELATIONAL_PATTERN.matcher(input).find()
                 || SPECIAL_CHAR_PATTERN.matcher(input).find();
     }
-
-    static FlsStripInaccessibleWarningInfo getFlsStripInaccessibleWarningInfo(
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/rules/fls/apex/operations/FlsViolationMessageUtil.java:232]: Moved these to FlsViolationCreatorUtil.java

File: sfge/src/main/java/com/salesforce/rules/fls/apex/operations/UnresolvedCrudFlsViolation.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * FLS Violation when SFGE understands that a DML operation is happening, but it is unable to
+ * determine more information about the object and the fields involved.
+ */
+public class UnresolvedCrudFlsViolation extends FlsViolationInfo {
+    public UnresolvedCrudFlsViolation(FlsConstants.FlsValidationType validationType) {
+        super(validationType, SoqlParserUtil.UNKNOWN, CollectionUtil.newTreeSet(), false);
+    }
+
+    public String getMessageTemplate() {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/rules/fls/apex/operations/UnresolvedCrudFlsViolation.java:16]: Message template is now owned by each violation info.

File: sfge/src/test/java/com/salesforce/testutils/ViolationWrapper.java [hunk 8]
```diff
@@ -2,15 +2,47 @@
...
-import com.salesforce.rules.fls.apex.operations.FlsViolationUtils;
+import com.salesforce.rules.fls.apex.operations.FlsViolationMessageUtil;
+import com.salesforce.rules.fls.apex.operations.UnresolvedCrudFlsViolation;
 import java.util.Arrays;
 import java.util.TreeSet;
+import java.util.function.Function;
 
 /** Wrapper around Violation to help comparing only the violation message and line numbers */
 public class ViolationWrapper {
+    public enum FlsViolationType {
```
Comment (rmohan20) [inline sfge/src/test/java/com/salesforce/testutils/ViolationWrapper.java:17]: Minor refactoring

**Review by jfeingold35 — APPROVED** [review]

**Pull request merged**